Add intervals query

romseygeek opened the pull request:
This commit exposes the lucene intervals query in elasticsearch, as a replacement for the 
Span query family.  Instead of building query structures up from individual terms, the intervals query
uses a tree of sources, the leaves of which are formed of `match`-type queries that are passed
through analysis.  These can then be combined using `or`, `combine` or `relate` sources to test
for proximity or position.

Replaces #32406, relates to #29636

elasticmachine commented:
Pinging @elastic/es-search

jimczi commented on rest-api-spec/src/main/resources/rest-api-spec/test/search/230_interval_query.yml:
```diff
@@ -0,0 +1,58 @@
+setup:
+  - skip:
```
Is this working ? I was not aware that the `skip` section can be set in the `setup` part but if it's working... ;)

jimczi commented on server/src/main/java/org/elasticsearch/index/query/IntervalBuilder.java:
```diff
@@ -0,0 +1,295 @@
+            if (posAtt.getPositionIncrement() == 1) {
+                if (synonyms.size() == 1) {
+                    terms.add(synonyms.get(0));
+                }
```
nit: can you add an `else` to desambiguate ?

jimczi commented on server/src/main/java/org/elasticsearch/index/query/IntervalQueryBuilder.java:
```diff
@@ -0,0 +1,111 @@
+import java.util.Objects;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+
```
Can you add javadocs to explain the kind of queries that this builder handles ?

jimczi commented on server/src/main/java/org/elasticsearch/index/query/IntervalsSourceProvider.java:
```diff
@@ -0,0 +1,490 @@
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
```
Same here, this is the user API so I'd expect some explanations regarding the different options ?

jimczi commented on server/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -811,12 +815,31 @@ private void registerQueryParsers(List<SearchPlugin> plugins) {
     }
 
+    private void registerIntervalsSourceProviders(List<SearchPlugin> plugins) {
+        registerIntervalsSourceProvider(new SearchPlugin.IntervalSpec<>(IntervalsSourceProvider.Match.NAME,
```
The `match` query is registered the same way so I wonder how the clash is handled internally ? We should maybe handle the source provider differently (e.g.: not as a named writeable) ?

romseygeek replied:
NamedWriteable is namespaced by class, so I think we're OK here

jimczi replied:
different [namespace](https://github.com/elastic/elasticsearch/pull/36135/files#diff-134bbc9191b667aff10076bf32f166adR837), got it, thanks.

romseygeek commented:
I've added some better YAML tests and some basic documentation in the query DSL reference.  I'm still a bit unsure of the DSL interface and some of the terms I'm using (eg `source` for something that generates intervals) but I think it's a good start. 

cc @clintongormley

mayya-sharipova commented on rest:
```diff
@@ -0,0 +1,5 @@
+
+./gradlew :distribution:archives:integ-test-zip:integTest \
+	-Dtests.class="org.elasticsearch.test.rest.*Yaml*IT" \
+	-Dtests.method="test {p0=$1}"
```
I wonder what is the role of this file?

romseygeek commented:
After conferring with @clintongormley I've reworked the API somewhat.  My one reservation is the name of the query, which is still `intervals`, which I think is still too based on the internal implementation of these queries.  Something like `proximity_match` might be better?  It links it to the other `match_*` queries.  I'd also like to rename the internal `intervals` field names to something like `rules`.

jpountz commented on docs/reference/query-dsl/intervals-query.asciidoc:
```diff
@@ -0,0 +1,155 @@
+favourite food is porridge` would not match, because the interval matching
+`cold porridge` starts before the interval matching `my favourite food`.
+
+[[intervals-match]]
```
Huge +1 to exposing a match and no term

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java:
```diff
@@ -579,6 +581,15 @@ public Query existsQuery(QueryShardContext context) {
+        @Override
+        public IntervalsSource intervals(String text, int maxGaps, boolean ordered, NamedAnalyzer analyzer) throws IOException {
+            if (indexOptions().compareTo(IndexOptions.DOCS_AND_FREQS_AND_POSITIONS) < 0) {
+                throw new IllegalArgumentException("Cannot create source against field [" + name() + "] with no positions indexed");
```
maybe say "intervals source" instead of just source?

romseygeek replied:
++

jpountz commented on server/src/main/java/org/elasticsearch/index/query/IntervalBuilder.java:
```diff
@@ -0,0 +1,299 @@
+/**
+ * Constructs an IntervalsSource based on analyzed text
+ */
+public class IntervalBuilder {
```
Can we make it pkg-private?

romseygeek replied:
It's called from TextFieldMapper which is in a different package, so unfortunately we can't.

jpountz replied:
should we move it to the same package to keep visibility to a minimum?

romseygeek replied:
It's also called from IntervalsSourceProvider, which doesn't really fit in the `mapper` package.  I think it has to stay public...

jpountz commented on server/src/main/java/org/elasticsearch/index/query/IntervalQueryBuilder.java:
```diff
@@ -0,0 +1,144 @@
+                default:
+                    provider = IntervalsSourceProvider.fromXContent(parser);
+
+            }
```
it seems that this won't fail if multiple providers are provided?

romseygeek replied:
Have updated

jpountz commented on server/src/main/java/org/elasticsearch/index/query/IntervalQueryBuilder.java:
```diff
@@ -0,0 +1,144 @@
+        MappedFieldType fieldType = context.fieldMapper(field);
+        if (fieldType == null) {
+            throw new IllegalArgumentException("Cannot create IntervalQuery over non-existent field [" + field + "]");
+        }
```
in general we are lenient with unmapped fields because of cross-index search, should we be lenient here too?

romseygeek replied:
Have updated

jpountz commented on docs/reference/query-dsl/intervals-query.asciidoc:
```diff
@@ -0,0 +1,155 @@
+The `or` rule will match any of its nested sub-rules.
+
+[horizontal]
+`intervals`::
```
Maybe we need to add documentation here about the fact that this only returns minimal intervals and consequences?

romseygeek replied:
I've added a section on minimization at the end of the doc, with some examples of queries that can produce surprising results, and how to deal with that.

jpountz replied:
awesome

romseygeek commented:
I've pushed some changes addressing your comments @jpountz.  For the filtering, it seemed to make more sense to add it as an option on to each rule, because the intervals produced by a filter all belong to the rule being filtered, and Clint and I thought that this was the best way of making that obvious.

romseygeek commented:
@elasticmachine retest this please

romseygeek commented:
@elasticmachine retest this please

clintongormley commented:
> Thanks @romseygeek. I thought having some sort of filtering intervals would be more consistent with how we deal with queries but I don't feel too strongly about it either.

@jpountz I didn't have strong reasons for adding filtering the way it is here.  Happy to discuss if you think it makes sense doing it in a different way.
